agent/stats/engine_unix_test.go

```diff
@@ -235,3 +247,176 @@ func TestFetchEBSVolumeMetrics(t *testing.T) {
+func TestFetchEBSVolumeMetricsNoTimeoutError(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	resolver := mock_resolver.NewMockContainerMetadataResolver(mockCtrl)
+	mockDockerClient := mock_dockerapi.NewMockDockerClient(mockCtrl)
+	t1 := &apitask.Task{
+		Arn: "t1",
+		Volumes: []apitask.TaskVolume{
+			{
+				Name: "1",
+				Type: apiresource.EBSTaskAttach,
+				Volume: &taskresourcevolume.EBSTaskVolumeConfig{
+					VolumeId:             "vol-12345",
+					VolumeName:           "test-volume",
+					VolumeSizeGib:        "10",
+					SourceVolumeHostPath: "taskarn_vol-12345",
+					DeviceName:           "/dev/nvme1n1",
+					FileSystem:           "ext4",
+				},
+			},
+		},
+	}
+
+	resolver.EXPECT().ResolveTaskByARN("t1").AnyTimes().Return(t1, nil)
+	mockDockerClient.EXPECT().Stats(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil).AnyTimes()
+
+	engine := NewDockerStatsEngine(&cfg, nil, eventStream("TestFetchEBSVolumeMetrics"), nil, nil)
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	engine.ctx = ctx
+	engine.resolver = resolver
+	engine.cluster = defaultCluster
+	engine.containerInstanceArn = defaultContainerInstance
+	engine.client = mockDockerClient
+
+	timeoutctx, timeoutcancel := context.WithTimeout(ctx, 1*time.Second)
+	defer timeoutcancel()
+	mockCsiClient := mock_csiclient.NewMockCSIClient(mockCtrl)
+	mockCsiClient.EXPECT().GetVolumeMetrics(gomock.Any(), "vol-12345", "/mnt/ecs/ebs/taskarn_vol-12345").Do(func(ctx context.Context, volID, hostPath string) {
+		time.Sleep(5 * time.Millisecond)
+	}).Return(&csiclient.Metrics{
+		Used:     15 * 1024 * 1024 * 1024,
+		Capacity: 20 * 1024 * 1024 * 1024,
+	}, nil).MinTimes(1)
+	engine.csiClient = mockCsiClient
+
+	expectedUsedBytes := aws.Float64(15 * 1024 * 1024 * 1024)
+	expectedTotalBytes := aws.Float64(20 * 1024 * 1024 * 1024)
+	expectedMetrics := []*ecstcs.VolumeMetric{
+		{
+			VolumeId:   aws.String("vol-12345"),
+			VolumeName: aws.String("test-volume"),
+			Utilized: &ecstcs.UDoubleCWStatsSet{
+				Max:         expectedUsedBytes,
+				Min:         expectedUsedBytes,
+				SampleCount: aws.Int64(1),
+				Sum:         expectedUsedBytes,
+			},
+			Size: &ecstcs.UDoubleCWStatsSet{
+				Max:         expectedTotalBytes,
+				Min:         expectedTotalBytes,
+				SampleCount: aws.Int64(1),
+				Sum:         expectedTotalBytes,
+			},
+		},
+	}
+
+	actualMetrics := engine.fetchEBSVolumeMetrics(t1, "t1")
+
+	require.NoError(t, timeoutctx.Err())
+	assert.Len(t, actualMetrics, 1)
+	assert.Equal(t, actualMetrics, expectedMetrics)
+}
+
+// TestFetchEBSVolumeMetricsTimeoutError tests that there was a timeout error when fetching volume metrics for an EBS volume via the CSI client. The expected behavior is
+// to have no volume metrics withn the output of fetchEBSVolumeMetrics and return with a context deadline exceeded error.
+func TestFetchEBSVolumeMetricsTimeoutError(t *testing.T) {
```

Comment:
nit: might want to use table-driven tests to reduce duplicate code.

Reply:
done, great suggestion!